mmdet3d/ops/paconv/paconv.py

```diff
@@ -0,0 +1,350 @@
+            Defaults to 'neighbor'.
+        scorenet_cfg (dict, optional): Config of the ScoreNet module, which
+            may contain the following keys and values:
+
+            - mlp_channels (List[int]): Hidden units of MLPs.
+            - score_norm (str): Normalization function of output scores.
+                Can be 'softmax', 'sigmoid' or 'identity'.
+            - temp_factor (float): Temperature factor to scale the output
+                scores before softmax.
+            - last_bn (bool): Whether to use BN on the last output of mlps.
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 out_channels: int,
+                 num_kernels: int,
+                 norm_cfg: dict = dict(type='BN2d', momentum=0.1),
+                 act_cfg: dict = dict(type='ReLU', inplace=True),
+                 scorenet_input: str = 'ed7',
+                 weight_bank_init: str = 'kaiming',
+                 kernel_input: str = 'neighbor',
+                 scorenet_cfg: dict = dict(
+                     mlp_channels=[8, 16, 16],
+                     score_norm='softmax',
+                     temp_factor=1.0,
+                     last_bn=False)):
+        super(PAConv, self).__init__()
+
+        # determine weight kernel size according to used features
+        if kernel_input == 'identity':
+            # only use grouped_features
+            self.kernel_mul = 1
+        elif kernel_input == 'neighbor':
+            # concat of (grouped_features - center_features, grouped_features)
+            self.kernel_mul = 2
+        else:
+            raise NotImplementedError(
+                f'unsupported kernel_input {kernel_input}')
+        self.kernel_input = kernel_input
+
+        # determine mlp channels in ScoreNet according to used xyz features
+        if scorenet_input == 'identity':
+            # only use relative position (grouped_xyz - center_xyz)
+            self.scorenet_in_channels = 3
+        elif scorenet_input == 'neighbor':
+            # (grouped_xyz - center_xyz, grouped_xyz)
+            self.scorenet_in_channels = 6
+        elif scorenet_input == 'ed7':
+            # (center_xyz, grouped_xyz - center_xyz, Euclidian distance)
+            self.scorenet_in_channels = 7
+        else:
+            raise NotImplementedError(
+                f'unsupported scorenet_input {scorenet_input}')
+        self.scorenet_input = scorenet_input
+
+        # construct weight kernels in weight bank
+        # self.weight_bank is of shape [C, num_kernels * out_c]
+        # where C can be in_c or (2 * in_c)
+        if weight_bank_init == 'kaiming':
+            weight_init = nn.init.kaiming_normal_
+        elif weight_bank_init == 'xavier':
+            weight_init = nn.init.xavier_normal_
+        else:
+            raise NotImplementedError(
+                f'unsupported weight bank init method {weight_bank_init}')
+
+        self.m = num_kernels
```

Comment:
`m` is so abstract? still use self.num_kernels would be better?

Reply:
Yes, I have changed this variable name in later [branch](https://github.com/Wuziyi616/mmdetection3d/blob/paconv_model/mmdet3d/ops/paconv/paconv.py#L187). Maybe we can just leave it as is and change in later merge of that branch?